### ISSUE TITLE
ENH: Add `__slots__ = ()` to `NDArrayOperatorsMixin`

### ISSUE THREAD

**Issue by fbunt** [numpy/numpy#22876]
### Proposed new feature or change:

Currently, if any parent class in an inheritance chain does not define `__slots__`, then all subclasses allow dynamic attribute creation. Even if every subsequent subclass defines `__slots__`, attributes are still able to be added to instances. 

It would be really nice if the `NDArrayOperatorsMixin` class defined `__slots__ = ()`. This would allow subclasses to prevent dynamic attribute creation by also using `__slots__`. I don't think there is a downside to adding this and the upside is that  subclasses could then choose to add slots of their own and prevent dynamic attribute creation.

**Comment by mattip** [comment]
Where did you run into this?

**Comment by fbunt** [comment]
I have an array-like class that implements the __array_ufunc__ interface. When I first made it, I overloaded all of the operator methods myself. I had `__slots__` defined on my class so that no more attributes could be added. I then discovered `NDArrayOperatorsMixin` and refactored my class to inherit from it in order to simplify the code. The code is nicer but I can't prevent more attributes being added without overloading `__setattr__` and checking if the given attribute name is in `__slots__`, `__dict__`, or `dir(self)` (the last one allows property setters).
